x-pack/plugins/observability_solution/entity_manager/server/lib/entities/install_entity_definition.ts

```diff
@@ -218,3 +188,123 @@ async function installAndStartDefinition(
   }
   return definition;
 }
+
+// perform installation of an entity definition including all
+// the necessary components
+async function install({
+  esClient,
+  soClient,
+  definition,
+  logger,
+}: InstallDefinitionParams): Promise<EntityDefinition> {
+  logger.debug(
+    () =>
+      `Installing definition ${definition.id} v${definition.version}\n${JSON.stringify(
+        definition,
+        null,
+        2
+      )}`
+  );
+
+  const entityDefinition = await saveEntityDefinition(soClient, {
+    ...definition,
+    installStatus: 'installing',
+    installStartedAt: new Date().toISOString(),
+  });
+
+  logger.debug(`Installing index templates for definition ${definition.id}`);
+  await Promise.allSettled([
+    upsertTemplate({
+      esClient,
+      logger,
+      template: getEntitiesHistoryIndexTemplateConfig(definition),
+    }),
+    upsertTemplate({
+      esClient,
+      logger,
+      template: getEntitiesLatestIndexTemplateConfig(definition),
+    }),
+  ]).then(throwIfRejected);
+
+  logger.debug(`Installing ingest pipelines for definition ${definition.id}`);
+  await Promise.allSettled([
+    createAndInstallHistoryIngestPipeline(esClient, definition, logger),
+    createAndInstallLatestIngestPipeline(esClient, definition, logger),
+  ]).then(throwIfRejected);
+
+  logger.debug(`Installing transforms for definition ${definition.id}`);
+  await Promise.allSettled([
+    createAndInstallHistoryTransform(esClient, definition, logger),
+    isBackfillEnabled(definition)
+      ? createAndInstallHistoryBackfillTransform(esClient, definition, logger)
+      : Promise.resolve(),
+    createAndInstallLatestTransform(esClient, definition, logger),
+  ]).then(throwIfRejected);
+
+  await updateEntityDefinition(soClient, definition.id, { installStatus: 'installed' });
+
+  return entityDefinition;
+}
+
+// stop and delete the current transforms and reinstall all the components
+async function reinstall({
+  esClient,
+  soClient,
+  definition,
+  latestDefinition,
+  logger,
+}: InstallDefinitionParams & { latestDefinition: EntityDefinition }): Promise<EntityDefinition> {
+  logger.debug(
+    `Reinstalling definition ${definition.id} from v${definition.version} to v${latestDefinition.version}`
+  );
+
+  try {
+    await updateEntityDefinition(soClient, latestDefinition.id, {
+      ...latestDefinition,
+      installStatus: 'installing',
```

Comment:
Should we have a dedicated status here for `updating`?
Makes it a bit harder to run the `shouldReinstall` check but might make it more clear for us if it failed on the first try or on an update

Reply:
good point, I think what I had in mind is that when doing an installation without prior state we'd remove everything including the saved object. It's only when installing a newer version that we leave this in a corrupted state so we already know if a definition is in an installing state it's during an upgrade. we can make that more explicit

Reply:
added in https://github.com/elastic/kibana/pull/188351/commits/591fc2011978f0aef2d2bf6bb238493e9cc8013e

Reply:
added in https://github.com/elastic/kibana/pull/188351/commits/591fc2011978f0aef2d2bf6bb238493e9cc8013e